File: python/src/kiln/cache.py
"""Content-addressed disk cache for fetched boundary attachments.

Admin boundaries change at most once a year; a country registry can carry
tens of thousands of externally-referenced ones. Caching the fetched bytes
by URL means every run after the first is near-instant, and gives
resumability as a side effect -- a run killed halfway does not re-fetch
what it already has.

Each cache entry is two files, both named after a sha256 hex digest of the
URL so an arbitrary URL always maps to a filesystem-safe name:

    <hash>.bin         the raw fetched bytes, exactly as received
    <hash>.meta.json   {"url": ..., "fetched_at": "<UTC ISO8601>", "sha256": "<hex digest of .bin>"}

The metadata sidecar exists so the cache directory is not a dead end for
debugging: a directory of bare hex-named files gives no way to tell what a
stale or wrong-looking entry came from. Given a suspicious boundary, `grep`
the `.meta.json` files for its URL to find the entry and see when it was
fetched. `sha256` isn't just for debugging, though -- it's checked on every
read (see `cache_read`) to catch a truncated or bit-flipped `.bin` file
before it's fed into the pipeline as if it were good data.

Deliberately NOT stored: response `Content-Type`. Storing it would mean
threading the `httpx.Response` (or at least its headers) through the
worker-thread return value into this module, for a field that's already
recoverable another way -- the resolved attachment's own `contentType` is
written into the output NDJSON right next to the base64 data, so it's not
lost, just not duplicated here.

Both files are written with a temp-file-then-`os.replace`, the same pattern
`write.py` uses to publish a partition (see `write.py::_finalize`): the
fetch path runs on a pool of worker threads, and two of them can race to
fill the same cache entry for the same URL. `os.replace` is atomic on both
POSIX and Windows, so a reader never observes a partially-written file. The
write order -- `.bin` before `.meta.json` -- makes the metadata file's
existence itself the commit marker: a reader that finds a `.meta.json`
alongside a `.bin` is guaranteed the `.bin` it names was fully written,
never a half-finished one from a write still in progress.

Nothing in this module raises. A cache is an optimization, not a
correctness requirement: an unwritable directory, a corrupt or truncated
entry, or any other OSError degrades to "no cache" for that entry (or
silently drops the write it can't make), reported back to the caller as an
error string rather than raised -- a full disk must not fail someone's
export. Callers (`extract.py`) are responsible for turning that string into
a `Report` entry; this module never touches `Report` itself; it runs on
worker threads, and `Report` is only ever mutated from the main thread (see
`extract.py::_resolve_boundary_work`) so report ordering stays
deterministic across runs.
"""

from __future__ import annotations

import hashlib
import json
import os
import tempfile
from datetime import datetime, timezone
from pathlib import Path

DEFAULT_CACHE_DIR = Path.home() / ".cache" / "kiln" / "boundaries"


def cache_key(url: str) -> str:
    """sha256 hex digest of `url`, used as the cache entry's filename stem."""
    return hashlib.sha256(url.encode()).hexdigest()


def _entry_paths(cache_dir: Path, url: str) -> tuple[Path, Path]:
    key = cache_key(url)
    return cache_dir / f"{key}.bin", cache_dir / f"{key}.meta.json"


def cache_read(cache_dir: Path, url: str) -> tuple[bytes | None, str | None]:
    """Return `(payload, error)` for `url`.

    A plain miss (no entry yet) is `(None, None)` -- not reported, this is
    the expected, common case for a first run. Anything else that keeps
    this from being a clean hit -- unreadable metadata, metadata that
    doesn't parse, a `.bin` file whose content doesn't match the recorded
    sha256 (truncated or corrupted), or an OSError reading either file --
    is also treated as a miss (`payload is None`, so the caller re-fetches)
    but comes back with `error` set so the caller can report it.
    """
    bin_path, meta_path = _entry_paths(cache_dir, url)

    try:
        meta_raw = meta_path.read_text(encoding="utf-8")
    except FileNotFoundError:
        return None, None
    except OSError as exc:
        return None, f"could not read cache metadata for {url}: {exc}"

    try:
        meta = json.loads(meta_raw)
        expected_sha = meta["sha256"]
    except (json.JSONDecodeError, KeyError, TypeError) as exc:
        return None, f"cache metadata for {url} is corrupt: {exc}"

    try:
        payload = bin_path.read_bytes()
    except OSError as exc:
        return None, f"could not read cache entry for {url}: {exc}"

    if hashlib.sha256(payload).hexdigest() != expected_sha:
        return None, f"cache entry for {url} failed its integrity check; re-fetching"

    return payload, None


def cache_write(cache_dir: Path, url: str, payload: bytes) -> str | None:
    """Write `payload` for `url`, best-effort. Returns an error string, or
    None on success.

    Callers must only call this after a successful fetch -- never cache a
    failure, since a 404 today may be a working URL tomorrow, and caching
    the failure would make a transient outage permanent.
    """
    bin_path, meta_path = _entry_paths(cache_dir, url)
    sha = hashlib.sha256(payload).hexdigest()
    meta = json.dumps(
        {
            "url": url,
            "fetched_at": datetime.now(timezone.utc).isoformat(),
            "sha256": sha,
        }
    ).encode("utf-8")

    try:
        cache_dir.mkdir(parents=True, exist_ok=True)
        _atomic_write(bin_path, payload)
        # Written last: its existence is the commit marker a reader relies
        # on to know the .bin next to it is complete (see module docstring).
        _atomic_write(meta_path, meta)
    except OSError as exc:
        return f"could not write cache entry for {url}: {exc}"

    return None


def _atomic_write(path: Path, data: bytes) -> None:
    """Write `data` to `path` via a same-directory temp file + `os.replace`.

    Same pattern as `write.py::_finalize`: the temp file lives in `path`'s
    own directory (so the final `os.replace` is guaranteed same-filesystem,
    hence atomic) and is only renamed into place once it holds the
    complete, correct content -- `path` itself is never opened for writing
    directly, which would let a concurrent reader (another worker thread
    racing on the same URL) observe a partial file.
    """
    fd, tmp_name = tempfile.mkstemp(dir=path.parent, prefix=f".{path.name}.", suffix=".tmp")
    try:
        with os.fdopen(fd, "wb") as handle:
            handle.write(data)
        os.replace(tmp_name, path)
    except BaseException:
        try:
            os.unlink(tmp_name)
        except OSError:
            pass
        raise

File: src/extract/cache.rs
//! Content addressed cache of fetched boundary bytes, keyed by sha256 of the
//! URL. Two files per entry: `<key>.bin` and `<key>.meta.json` (url,
//! fetched_at, sha256 of the bin, verified on read). Written temp-then-rename,
//! bin before meta, so a meta file is the commit marker. Nothing here aborts a
//! run: every failure is an `Err(String)` the caller reports as `cache_error`.
//! Same layout as the Python cache, so directories are interchangeable.

use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicU64, Ordering};

use sha2::{Digest, Sha256};

use crate::snapshot::instant::format_utc;

static TMP_SEQ: AtomicU64 = AtomicU64::new(0);

pub fn cache_key(url: &str) -> String {
    hex(&Sha256::digest(url.as_bytes()))
}

fn hex(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

#[derive(Debug, Clone)]
pub struct Cache {
    dir: PathBuf,
}

impl Cache {
    pub fn new(dir: &Path) -> Self {
        Self {
            dir: dir.to_path_buf(),
        }
    }

    fn paths(&self, url: &str) -> (PathBuf, PathBuf) {
        let key = cache_key(url);
        (
            self.dir.join(format!("{key}.bin")),
            self.dir.join(format!("{key}.meta.json")),
        )
    }

    /// Ok(None) on a miss (no meta file). Err on a corrupt or unreadable entry.
    /// Treat an Err the same as a miss -- re-fetch the URL -- and report it
    /// as `cache_error`; never serve the entry it names as good data.
    pub fn read(&self, url: &str) -> std::result::Result<Option<Vec<u8>>, String> {
        let (bin, meta) = self.paths(url);
        let meta_text = match std::fs::read_to_string(&meta) {
            Ok(t) => t,
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(None),
            Err(e) => return Err(format!("{}: {e}", meta.display())),
        };
        let meta_json: serde_json::Value =
            serde_json::from_str(&meta_text).map_err(|e| format!("{}: {e}", meta.display()))?;
        let expected = meta_json["sha256"]
            .as_str()
            .ok_or_else(|| format!("{}: no sha256", meta.display()))?;
        let bytes = std::fs::read(&bin).map_err(|e| format!("{}: {e}", bin.display()))?;
        let actual = hex(&Sha256::digest(&bytes));
        if actual != expected {
            return Err(format!(
                "{}: sha256 mismatch (entry corrupt or truncated)",
                bin.display()
            ));
        }
        Ok(Some(bytes))
    }

    /// Two writers racing on the same URL with differing bytes can leave a
    /// mismatched bin/meta pair (one writer's bin next to the other's meta);
    /// `read` detects that via the sha256 check and reports it as an error
    /// rather than serving either writer's bytes as if they were the other's.
    pub fn write(&self, url: &str, bytes: &[u8]) -> std::result::Result<(), String> {
        std::fs::create_dir_all(&self.dir).map_err(|e| format!("{}: {e}", self.dir.display()))?;
        let (bin, meta) = self.paths(url);
        atomic_write(&bin, bytes)?;
        let meta_json = serde_json::json!({
            "url": url,
            "fetched_at": format_utc(std::time::SystemTime::now()),
            "sha256": hex(&Sha256::digest(bytes)),
        });
        atomic_write(&meta, meta_json.to_string().as_bytes())
    }
}

/// Write to a sibling temp file unique to this process and call, then rename.
/// Two workers racing on the same entry each rename identical bytes; last wins.
fn atomic_write(path: &Path, bytes: &[u8]) -> std::result::Result<(), String> {
    let seq = TMP_SEQ.fetch_add(1, Ordering::Relaxed);
    let name = path
        .file_name()
        .map(|n| n.to_string_lossy().to_string())
        .unwrap_or_default();
    let tmp = path.with_file_name(format!(".{name}.{}-{seq}.tmp", std::process::id()));
    std::fs::write(&tmp, bytes).map_err(|e| format!("{}: {e}", tmp.display()))?;
    if let Err(e) = std::fs::rename(&tmp, path) {
        let _ = std::fs::remove_file(&tmp);
        return Err(format!("{}: {e}", path.display()));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn key_is_sha256_of_the_url_and_matches_python() {
        // python3 -c "import hashlib;print(hashlib.sha256(b'https://a/b').hexdigest())"
        let expected = "4148a87e16a32df5f7bd99effa01d69d8371d4e077fa2fca9f1cfca818ae02f8";
        assert_eq!(cache_key("https://a/b"), expected);
        assert_ne!(cache_key("https://a/b"), cache_key("https://a/c"));
    }

    #[test]
    fn write_then_read_round_trips_and_leaves_no_temp_files() {
        let dir = tempfile::tempdir().unwrap();
        let cache = Cache::new(dir.path());
        cache.write("https://a/b", b"geo").unwrap();
        assert_eq!(cache.read("https://a/b").unwrap(), Some(b"geo".to_vec()));
        let names: Vec<String> = std::fs::read_dir(dir.path())
            .unwrap()
            .map(|e| e.unwrap().file_name().to_string_lossy().to_string())
            .collect();
        assert_eq!(names.len(), 2, "{names:?}");
        assert!(
            names.iter().any(|n| n.ends_with(".bin"))
                && names.iter().any(|n| n.ends_with(".meta.json"))
        );
        let meta: serde_json::Value = serde_json::from_str(
            &std::fs::read_to_string(
                dir.path()
                    .join(format!("{}.meta.json", cache_key("https://a/b"))),
            )
            .unwrap(),
        )
        .unwrap();
        assert_eq!(meta["url"], "https://a/b");
        assert!(meta["fetched_at"].as_str().unwrap().ends_with('Z'));
        assert_eq!(meta["sha256"].as_str().unwrap().len(), 64);
    }

    #[test]
    fn a_miss_is_none_and_a_corrupt_entry_is_an_error() {
        let dir = tempfile::tempdir().unwrap();
        let cache = Cache::new(dir.path());
        assert_eq!(cache.read("https://a/none").unwrap(), None);
        cache.write("https://a/b", b"geo").unwrap();
        std::fs::write(
            dir.path().join(format!("{}.bin", cache_key("https://a/b"))),
            b"tampered",
        )
        .unwrap();
        let err = cache.read("https://a/b").unwrap_err();
        assert!(err.contains("sha256"), "{err}");
    }

    #[test]
    fn a_bin_without_meta_is_a_miss_and_a_meta_without_bin_is_an_error() {
        let dir = tempfile::tempdir().unwrap();
        let cache = Cache::new(dir.path());
        std::fs::write(dir.path().join(format!("{}.bin", cache_key("u"))), b"x").unwrap();
        assert_eq!(cache.read("u").unwrap(), None);
        cache.write("v", b"y").unwrap();
        std::fs::remove_file(dir.path().join(format!("{}.bin", cache_key("v")))).unwrap();
        assert!(cache.read("v").is_err());
    }

    #[test]
    fn unwritable_dir_is_an_error_string_not_a_panic() {
        let f = tempfile::NamedTempFile::new().unwrap();
        // A file where a directory is needed.
        let cache = Cache::new(&f.path().join("sub"));
        assert!(cache.write("u", b"x").is_err());
    }

    #[test]
    fn concurrent_writes_of_different_bytes_to_the_same_url_are_never_torn() {
        let dir = tempfile::tempdir().unwrap();
        let cache = Cache::new(dir.path());
        let payloads: Vec<Vec<u8>> = (0..8u8).map(|i| vec![i; 4096]).collect();
        std::thread::scope(|s| {
            for payload in &payloads {
                let cache = cache.clone();
                s.spawn(move || cache.write("https://a/same", payload).unwrap());
            }
        });
        match cache.read("https://a/same") {
            Ok(Some(bytes)) => assert!(
                payloads.contains(&bytes),
                "read bytes did not match any single writer's payload"
            ),
            Err(e) => assert!(e.contains("sha256"), "{e}"),
            Ok(None) => panic!("expected a hit or a sha256 error, got a miss"),
        }
        assert_eq!(
            std::fs::read_dir(dir.path()).unwrap().count(),
            2,
            "no temp files left"
        );
    }
}
